Fix IPC4 pipeline trigger

ranj063 opened the pull request:
tested with https://github.com/thesofproject/sof/pull/6401

FIrst patch is from the previous PR #3914

RanderWang commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -32,47 +32,123 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
+	 * trigger each pipeline separately and if any of the pipeline trigger fails, its state
+	 * will be reset in hw_free, so no need to rollback states here
+	 */
+	for (i = 0; i < pipeline_info.count; i++) {
```
current sequence:
(1） pipeline 1 pause -> run, pipeline 2 pause -> run,  pipeline n pause -> run.
can we set multiple pipelines in one ipc msg like
(1) pipeline 1 ... n pause
(2) pipeline 1 ... n run

in current cSOF we deal with it like
````c
	if (state.extension.r.multi_ppl) {
		ppl_count = ppl_data->pipelines_count;
		ppl_id = ppl_data->ppl_id;
		dcache_invalidate_region((__sparse_force void __sparse_cache *)ppl_id,
					 sizeof(int) * ppl_count);
	} else {
		ppl_count = 1;
		id = state.primary.r.ppl_id;
		ppl_id = &id;
	}
````

ranj063 replied:
@RanderWang I used multiple pipeline IPC trigger and ran into all kinds of IPC errors with it. Thats why I updated this PR to remove the multiple pipeline IPC and replaced it with separate IPCs

ranj063 replied:
actually I take that back @RanderWang. I messed up the pipeline data for multiple pipelines. Fixed now.

RanderWang commented on sound/soc/sof/ipc4-topology.h:
```diff
@@ -82,6 +84,26 @@ struct sof_ipc4_pipeline {
+ */
+struct ipc4_pipeline_set_state_data {
+	uint32_t count;
+	uint32_t pipeline_ids[SOF_IPC4_TRIGGER_MAX_PIPELINE_COUNT];
```
struct sof_ipc4_multi_pipeline_info  is very similar to struct ipc4_pipeline_set_state_data.  can we use the same order for "uint32_t count" ?

RanderWang commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -32,65 +60,184 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
 
+/* helper function to get the list of pipeline widgets that need to be triggered */
+static int
+sof_ipc4_get_pipelines_in_order(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget *widget,
```
I don't understand how does it work since I am not good at recursion.  How about feedback stream like echo reference ?

As I know another method to sort pipelines is based on pipeline priority. Each pipeline has priority defined in topology and driver can sort pipeline with priority. But currently we don't set it in ipc4-topology.c

````c
	/* TODO: Get priority from topology */
	pipeline->priority = 0;
````

ranj063 replied:
priority has a different meaning that order @RanderWang. We may need to set the priority at some point later on to pass hints to the scheduler. We cannot mix up these 2 things

RanderWang replied:
I found windows driver use such priority to sort pipelines.

plbossart replied:
the priority is only meaningful to sort pipelines executed in LL context. It is not used for the data domain @RanderWang

bardliao commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -32,65 +60,184 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
 	struct snd_sof_pcm *spcm;
 	int ret = 0;
-	int num_widgets;
+	int num_widgets, i;
```
nitpick: Can we use single line for `int i;`?

RanderWang commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -32,65 +60,184 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
+		 * DAI type and for capture, it is of AIF type.
+		 */
+		if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK && !WIDGET_IS_DAI(widget->id))
+			continue;
```
in two streams mixed into to one case, the second stream will not have dai pipeline, only part of host to mixer

ranj063 replied:
@RanderWang that is incorrect. The list of dapm widgets for both pcm's will include the DAI widget

RanderWang replied:
if it is included and it is already running, will it be started again ?

RanderWang replied:
you can test it with current sdw or hda tplg, two streams:  deepbuffer and normal jack

ranj063 replied:
@RanderWang the firmware should be able ignore the state set up if it is already running

RanderWang commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -466,18 +468,28 @@ static int ipc4_hda_dai_trigger(struct snd_pcm_substream *substream,
 	case SNDRV_PCM_TRIGGER_START:
 	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
 		snd_hdac_ext_stream_start(hext_stream);
+		ret = sof_ipc4_set_pipeline_state(sdev, pipe_widget->instance_id,
```
do we need to paused it for PAUSE_RELEASE case ?  The stream is in PAUSE_PUSH state and already in paused state

ranj063 replied:
fixed now @RanderWang

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -58,24 +58,23 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 		if (!swidget)
 			continue;
 
+		/* find pipeline widget for the pipeline that this widget belongs to */
```
I know, this is copied, but what are we 'finding'?
Perhaps: "take the pipeline widget this widget belongs to" or just drop the comment?

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -58,24 +58,23 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 		case SOF_IPC4_PIPE_PAUSED:
 		case SOF_IPC4_PIPE_RESET:
-			if (!WIDGET_IS_AIF(swidget->id))
+			if (pipeline->is_backend)
```
The commit message implies that we did not skipped the BE triggering before, but we did, don't wee?
We only handled the AIF widget for PAUSE and RESET, so we did skipped BE...

ranj063 replied:
ive changed the message now.

ujfalusi commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1480,6 +1480,10 @@ static int sof_ipc4_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget
 
+		/* set the pipeline that the DAI widget belongs to as a backend pipeline */
+		pipeline = pipe_widget->private;
+		pipeline->is_backend = true;
```
Anticipating a 'yes, that's correct' answer:
I guess it is forbidden to have a pipeline directly connecting AIF to DAI, we must always have at least two pipelines running between AIF and DAI?

ranj063 replied:
Yes, thats correct

ujfalusi replied:
Chain DMA is not going to be like that?

ranj063 replied:
we will not be using the pipeline trigger IPC for chained DMA. so it doesnt matter

ujfalusi commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -466,18 +468,28 @@ static int ipc4_hda_dai_trigger(struct snd_pcm_substream *substream,
 	w = snd_soc_dai_get_widget(dai, substream->stream);
 	swidget = w->dobj.private;
+	pipe_widget = swidget->pipe_widget;
+	pipeline = pipe_widget->private;
```
I guess no need to do a safety check that `pipeline->is_backend == true`, right?

ranj063 replied:
not needed because we can only ever have a DAI widget set as the playback/capture widget for the BE DAIs

ujfalusi commented on sound/soc/sof/ipc4-topology.h:
```diff
@@ -82,6 +84,26 @@ struct sof_ipc4_pipeline {
+ * @count: Number of pipelines to be triggered
+ * @pipeline_ids: Arrays of IDs of the pipelines to be triggered
+ */
+struct ipc4_pipeline_set_state_data {
```
these are sent to the firmware, so need to be packed?

ranj063 replied:
yes, thanks for catching that

ujfalusi replied:
~~How can a compiler pack a flexible struct?~~

ujfalusi commented on include/sound/sof/ipc4/header.h:
```diff
@@ -185,6 +185,9 @@ enum sof_ipc4_pipeline_state {
 #define SOF_IPC4_GLB_PIPE_STATE(x)		((x) << SOF_IPC4_GLB_PIPE_STATE_SHIFT)
 
+#define SOF_IPC4_GLB_PIPE_MULTI			BIT(0)
+#define SOF_IPC4_GLB_PIPE_MULTI_SYNC_STOP	BIT(1)
```
SOF_IPC4_GLB_PIPE_MULTI_SYNC_STOP is for what?

so, this is a batch pipeline trigger feature?

ranj063 replied:
it is also needed for multiple pipeline trigger but when & for what is an open for now. But I've added it as it is part of the multi-pipeline trigger IPC bits

plbossart replied:
Erm, what is the open? That statement doesn't give me a warm and fuzzy feeling we should merge this. This looks like we're going to have multiple revisions if there are open conceptual issues on how this multi pipe stuff works.

ranj063 replied:
i dont think there's anyting conceptual missing here @plbossart. SYNC_STOP is a feature with multi-pipeline trigger that we just dont need right now

plbossart replied:
I find it confusing to add references to a TBD feature that we don't need. Let's add it if and when we have a need and actually use it.

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -32,65 +60,184 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
+		 * walking all the way to the source. So the starting widget for playback is of
+		 * DAI type and for capture, it is of AIF type.
+		 */
+		if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK && !WIDGET_IS_DAI(widget->id))
```
If we have a playback to more than one DAI?
```
aif --p1--> mix --p2-->DAI1
             |
             - --p3--> DAI2
```
You would collect: p2, p1, p3 (you skip the existing p1, afaik), but should not be p2, p3, p1 ?

ujfalusi replied:
Oh, you skip if the pipeline is DAI, so you will have a single p1 in the list?

ranj063 replied:
thats right

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -32,65 +60,184 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
+		return ret;
+
+	/* trigger a single pipeline if count is 1 */
+	if (pipeline_info.count == 1) {
```
and that single pipeline is the FE pipeline?

ranj063 replied:
yes

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -32,65 +60,184 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
+	for (i = 0; i < pipeline_info.count; i++) {
+		pipeline_widget = pipeline_info.pipelines[i];
+		pipeline = pipeline_widget->private;
+		pipeline->state = state;
```
So, we still have the separate BE (DAI) pipeline trigger and we handle the non BE (DAI) pipelines in batch mode?

ranj063 replied:
no if there are multiple BE DAI's, their pipelines will get triggered in the respective BE trigger callbacks. There wouldnt be a batch mode there

ujfalusi replied:
In here it does not matter if we have triggered it or not like in case of the pause?

ranj063 replied:
it does matter and if it is triggered it will be in the right state already no?

ranj063 replied:
I meant the ones newly triggered will need an update to the state and the others will be the same state already. So, no need to check here

ujfalusi commented:
@ranj063, I have started internal IPC4 daily test plan: 16213

ranj063 commented:
@RanderWang @ujfalusi @bardliao addressed the comments above. Please have a look again. Thanks!

RanderWang reviewed (approved):
LGTM

ujfalusi reviewed (changes requested):
@ranj063, the IPC4 test looks like a wreck with this PR...
internal run ID: 16257

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -13,6 +13,46 @@
+		if (!pipeline->is_backend) {
+			data.pipeline_ids[i] = info->pipelines[i]->instance_id;
+			data.count++;
+		}
```
@ranj063, the issue is that you record _all_ pipelines, including the BE one to the `pipeline_info` and when you have two pipelines in the route (one FW and one BE, capture) you would end up doing:
```
[ 8443.720222] sof_ipc4_set_multi_pipeline_state: info->count: 2
[ 8443.720226] sof_ipc4_set_multi_pipeline_state: i: 1, data.count: 1 (is_backend: 0)
[ 8443.720230] sof_ipc4_set_multi_pipeline_state: i: 0, data.count: 1 (is_backend: 1)
[ 8443.720234] sof_ipc4_set_multi_pipeline_state: data.count: 1
```
Put the FE pipeline at `data.pipeline_ids[1]` and leave the `data.pipeline_ids[0]` untouched, but you still tell the firmware that there is one pipeline to multi-set-state...

In `sof_set_up_widgets_in_path()` the comment say that the pipelines are added to the list in source to sink order, right?
Here you want to revert that order, if I read the comment correctly.
This diff fixes the issue, but I'm not sure if this is what you want:
```diff
diff --git a/sound/soc/sof/ipc4-pcm.c b/sound/soc/sof/ipc4-pcm.c
index 85a96b75ebb7..b20c61a02d2b 100644
--- a/sound/soc/sof/ipc4-pcm.c
+++ b/sound/soc/sof/ipc4-pcm.c
@@ -40,10 +40,8 @@ static int sof_ipc4_set_multi_pipeline_state(struct snd_sof_dev *sdev, u32 state
 		struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
 
 		/* skip triggering backend widgets */
-		if (!pipeline->is_backend) {
-			data.pipeline_ids[i] = info->pipelines[i]->instance_id;
-			data.count++;
-		}
+		if (!pipeline->is_backend)
+			data.pipeline_ids[data.count++] = info->pipelines[i]->instance_id;
 	}
 
 	ipc_size = sizeof(u32) * (data.count + 1);

```

ranj063 replied:
@ujfalusi I messed this one royally when I switched the order of iterating over the array. fixed now. Thanks!

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -13,6 +13,46 @@
+		}
+	}
+
+	ipc_size = sizeof(u32) * (data.count + 1);
```
It took me a while to figure out what that `+ 1` is doing there, hint: it is for the size of the `data.count`

ranj063 replied:
added a comment now

ujfalusi commented:
Internal IPC4 daily plan started for the PR: 16339

ujfalusi commented:
> Internal IPC4 daily plan started for the PR: 16339

Looks like all playback tests on TGLU_RVP_NOCODEC_IPC4ZPH fails with this PR...

ujfalusi commented:
@ranj063, I can reproduce the same playback failure locally. Rebased your branch and did a bisect:
```
1b0ed6c99128c8db6106c1f3b9c75214a0e66ba2 is the first bad commit
commit 1b0ed6c99128c8db6106c1f3b9c75214a0e66ba2
Author: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>
Date:   Tue Oct 11 10:59:03 2022 -0700

    ASoC: SOF: IPC4: move BE pipeline trigger in the BE DAI trigger
    
    Move all BE pipeline triggers to the BE DAI trigger including the ones
    for START and PAUSE_RELEASE.
    
    Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

 sound/soc/sof/intel/hda-dai.c | 39 +++++++++++++++++++++++++++++++++------
 sound/soc/sof/ipc4-pcm.c      | 16 ++--------------
 2 files changed, 35 insertions(+), 20 deletions(-)

$ git bisect log
git bisect start
# status: waiting for both good and bad commits
# good: [ea1ca7d0610c8cc4cd8101e239694828af20da7e] ASoC: SOF: introduce new DEBUG_NOCODEC mode
git bisect good ea1ca7d0610c8cc4cd8101e239694828af20da7e
# status: waiting for bad commit, 1 good commit known
# bad: [b67f9b8181175a0b6d307f25f68b4f3c7fdaa1f7] ASoC: SOF: ipc4: Add support for triggering pipelines in order
git bisect bad b67f9b8181175a0b6d307f25f68b4f3c7fdaa1f7
# bad: [1b0ed6c99128c8db6106c1f3b9c75214a0e66ba2] ASoC: SOF: IPC4: move BE pipeline trigger in the BE DAI trigger
git bisect bad 1b0ed6c99128c8db6106c1f3b9c75214a0e66ba2
# good: [0978d1dbd6cdea8b6af2b4bb61c4fd52cc656f31] ASoC: SOF: ipc4-pcm: Fix skipping backend pipelines during FE trigger
git bisect good 0978d1dbd6cdea8b6af2b4bb61c4fd52cc656f31
# first bad commit: [1b0ed6c99128c8db6106c1f3b9c75214a0e66ba2] ASoC: SOF: IPC4: move BE pipeline trigger in the BE DAI trigger
```

ujfalusi reviewed (changes requested):
NOCODEC with SOC.IPC4 does not like the audio playback. Might be firmware issue, but I can reproduce locally.

bardliao commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -13,6 +13,33 @@
+
+	/* trigger a single pipeline */
+	if (data->count == 1)
+		return sof_ipc4_set_pipeline_state(sdev, data->pipeline_ids[0], state);
```
Can we move this to the beginning of this function?

RanderWang commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -37,73 +64,84 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 
-		pipeline->state = state;
+		for (j = 0; j < data.count; j++) {
+			if (data.pipeline_ids[j] == pipeline_info->pipelines[i]->instance_id) {
```
you don't need to check this here since if a pipeline is not in data.pipeline_ids,  pipeline->state should be equal to state in above 
````c
		if (pipeline->state != state)
			data.pipeline_ids[data.count++] = pipeline_info->pipelines[i]->instance_id;
````

RanderWang reviewed (approved):
LGTM

libinyang commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -13,6 +13,33 @@
+{
+	struct sof_ipc4_msg msg = {{ 0 }};
+	u32 primary, ipc_size;
+
```
Need to check ```count == 0```

ujfalusi replied:
It is checked in sof_ipc4_set_pipeline_state()

libinyang commented on sound/soc/sof/sof-audio.h:
```diff
@@ -46,6 +46,8 @@
 #define SOF_DAI_CLK_INTEL_SSP_MCLK	0
 #define SOF_DAI_CLK_INTEL_SSP_BCLK	1
 
+#define SOF_TRIGGER_MAX_PIPELINE_COUNT 16
```
What the limitation of ```16``` comes from?

ranj063 replied:
thats the max number of pipelines in the FW

plbossart replied:
The restriction is on the number of pipelines that have a window for timestamping. We have no fixed restriction to 16 in general, and if you see the number of DMAs we are past 16 already.

@bardliao added something to read the actual number of pipelines from the firmware, can you refresh everyone's memory?

ranj063 replied:
@plbossart that's correct but I need to create a fixed-size array for IPC playload here and I need a number. So I pixked 16 because it seemed like a large enough number for connected pipelines

plbossart replied:
why not allocate dynamically based on what the firmware actually supports?

ranj063 replied:
@plbossart this is exactly how I started did this PR first and remember I told you the multi-pipeline IPC always fails with both the cavs and the SOF firmware? this is because I couldnt use dynamically allocated data with the IPC

plbossart replied:
erm, what? we allocate such memory all the time. Not getting the issue, sorry.

ranj063 replied:
@plbossart could you please point me to an IPC structure containing a dynamically allocated data pointer?

plbossart replied:
you can allocate a u8 buffer and cast to a structure as needed?

ranj063 replied:
@plbossart im not sure I follow the suggestion. In the end the the pipeline_ids[] array in struct ipc4_pipeline_set_state_data has to be of fixed size right?

ranj063 replied:
or are you asking me to make it a flex array?

plbossart replied:
isn't that what the 'count' is for, to tell the firmware how many pipelines to look for?

ranj063 replied:
but the count is not fixed isnt it? It depends on how many pipelines that PCM has

ranj063 replied:
let me rework the PR with a flex array. Maybe then I can removethe max count restriction

bardliao replied:
> The restriction is on the number of pipelines that have a window for timestamping. We have no fixed restriction to 16 in general, and if you see the number of DMAs we are past 16 already.
> 
> @bardliao added something to read the actual number of pipelines from the firmware, can you refresh everyone's memory?

https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/ipc4-loader.c#L386 in `sof_ipc4_query_fw_configuration`. But I think @ranj063 already know it. :)

libinyang commented on sound/soc/sof/sof-audio.c:
```diff
@@ -394,15 +395,41 @@ static int sof_free_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dap
+				break;
+
+		if (i == pipeline_info->count) {
+			u32 pipeline_count = pipeline_info->count++;
```
No need to define a new variable. Can use ```i``` directly.

ranj063 replied:
yes, let me do this in a fixup

libinyang commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -37,72 +64,77 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
-		default:
-			break;
-		}
+		if (pipeline->state != state)
```
As next state it is setting is ```SOF_IPC4_PIPE_PAUSED```, we should not skip ```pipeline->state == state``` pipelines.

kv2019i replied:
@libinyang but the PAUSE is just an intermediate state, so we can skip the change if pipeline is already in target state. otherwise we'll do an unnessary STATEFOO->PAUSED->STATEFOO if function is called to set state to STATEFOO. Right?

ranj063 replied:
@kv2019i yes thats correct, it will result in failures if we checks for PAUSED because with mixer-based pipelines, the BE pipeline will be RUNNING, not paused

libinyang replied:
Sorry, I reopen my concern here. For a single pipeline, it may be OK. But for multiple pipelines I'm not sure it's OK or not. For example, ppl1 -> ppl2 -> ppl3. We want to make them into running state and ppl2 is already in running state (Not sure this assumption is valid or not). Then we pause ppl1 and ppl3, but we don't pause ppl2 with our code. So the ppl2 is still active in this case. Is this what we want?

ranj063 replied:
@libinyang you're talking a hypothetical scenario. How is it possible that you have pipeline 2 running while both pipeline 1 and 3 are paused. Wouldnt that pipeline immediately run into an xrun?

libinyang replied:
@ranj063 OK. If there is no such scenario, it's OK to me.

ujfalusi commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -466,18 +468,30 @@ static int ipc4_hda_dai_trigger(struct snd_pcm_substream *substream,
 	case SNDRV_PCM_TRIGGER_START:
 	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
 		snd_hdac_ext_stream_start(hext_stream);
+		if (pipeline->state != SOF_IPC4_PIPE_PAUSED) {
```
Is there a chance that the pipeline is in RUNNING state?
A second FE connects to the already running path via a mixer for example?
Will a pause/run toggle going to affect the already running stream?

ranj063 replied:
@ujfalusi no this will not and the BE should keep running as we handle this in the soc core. Pierre fixed this part and it works well even with IPC3

plbossart replied:
Humm, not so sure @ranj063 . What I did is prevent the triggers from happening twice at the BE dailink level. 

This is the dai level. We should double-check that indeed this is not going to happen.

ranj063 replied:
@plbossart Im 100% sure this works as expected. Please take a look at https://github.com/thesofproject/linux/blob/ea1ca7d0610c8cc4cd8101e239694828af20da7e/sound/soc/soc-pcm.c#L2118

plbossart replied:
yes indeed. the PCM is a level higher than both dailink and dai.

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -37,72 +64,77 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
-			dev_err(sdev->dev, "failed to set state %d for pipeline %d\n",
-				state, swidget->pipeline_id);
-			break;
+		for (j = 0; j < data.count; j++) {
```
Based on the comment on front of the topmost for() loop, it would be logical to use the `data.count` as the first loop (the pipelines that got triggered) and the inner loop to use `pipeline_info->count`

ujfalusi replied:
Or just reword the comment, so we will not need to reach down to `pipeline` in every inner loop?

ranj063 replied:
@ujfalusi can this be a fixup? I dont want to lose Rander/Bard's approval

kv2019i reviewed (commented):
I'll leave other to review the details in driver context, but overall approach seems good and use of "batched triggers" seems to be key for predictable results with complex tplgs. Otherwise FW will have varied timing between trigger actions on different parts of connected DSP dataflows (trying to avoid using "pipelines" as that has other meanings) and this has been demonstrated to result in hard-to-debug issues.

plbossart commented on sound/soc/sof/sof-audio.h:
```diff
@@ -290,6 +302,7 @@ struct snd_sof_pcm_stream {
 	 * active or not while suspending the stream
 	 */
 	bool suspend_ignored;
+	struct snd_sof_pipeline_trigger_info pipeline_info;
```
why is this here? This is IPC4 specific-stuff that pops at this level?

Shouldn't the information on connected pipelines be stored at the pipeline level?
Also not clear what we mean by 'connected'. If there is a switch between two pipelines, does this disconnect them?

ranj063 replied:
nothing IPC4 specific here at all. Its just a list of pipelines associated with a running stream isnt it? Its used by IPC4 and ignored by IPC3. but how can I store information about connected pipelines at the pipeline level? A pipeline is simply a widget right?

plbossart replied:
doesn't a widget have private information that can be used to store such information?

I get the argument that it doesn't hurt IPC3, but then we should also make an effort to avoid exposing new things that are IPC specific if we can.

ranj063 replied:
widget private is already used for its IPC struct payload. But even then, its not a widget properly right? i still would argue that the pipeline connections are a PCM stream property. 
And the main reason for doing this here in sof-audio (which is IPC-specific) is to avoid walking through the widgets once more to determine the trigger order when we can reuse the existing loop

ranj063 commented:
@RanderWang @bardliao @ujfalusi Do you mind taking a look again. I changed the PR to use flexible array for the set pipeline data and remove the max pipeline restriction and while I was at it, I also fixed the problem I saw with the stop trigger for HDA BE pipelines not getting skipped in the FE DAI trigger.

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -37,75 +64,92 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
+	list_for_each_entry(pipe_widget, &pipeline_info->pipeline_list, pcm_trigger_list) {
+		pipeline = pipe_widget->private;
+		if (pipeline->state != state && !pipeline->skip_fe_trigger)
+			data->pipeline_ids[i++] = pipe_widget->instance_id;
```
How does this enforce the 'sink to source' order that's described above?

ranj063 replied:
when adding the pipelines, they're added source->sink. So traversing the array in reverse ensures the needed order

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1801,6 +1801,7 @@ static int sof_ipc4_dai_config(struct snd_sof_dev *sdev, struct snd_sof_widget *
 	case SOF_DAI_INTEL_HDA:
 		gtw_attr = ipc4_copier->gtw_attr;
 		gtw_attr->lp_buffer_alloc = pipeline->lp_mode;
+		pipeline->skip_fe_trigger = true;
```
skip_during_fe_trigger?

I just don't see how a BE can skip an FE, all DPCM actions are initiated by FEs.

Also what is so special about HDaudio that requires a change here?

plbossart commented on sound/soc/sof/ipc4-topology.h:
```diff
@@ -71,15 +72,27 @@
  * @mem_usage: Memory usage
  * @state: Pipeline state
  * @msg: message structure for pipeline
+ * @skip_fe_trigger: skip triggering this pipeline in the FE DAI trigger
```
yeah, well. skip_fe_trigger doesn't mean skip during the FE trigger. Confusion between cause and effect I am afraid.

plbossart reviewed (commented):
this is not a simple PR to review. 

I don't know what to do here, approving means we are signing-up for fixups and not approving means I am the only black sheep in the team preventing CI from working.

ranj063 commented on sound/soc/sof/sof-audio.c:
```diff
@@ -394,29 +396,59 @@ static int sof_free_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dap
+				break;
+		}
+
+		dev_dbg(sdev->dev, "adding %s\n", swidget->pipe_widget->widget->name);
```
Note to self: Need to remove both dev_dbg prints here.

plbossart commented on sound/soc/sof/sof-audio.h:
```diff
@@ -106,6 +109,8 @@ struct sof_ipc_pcm_ops {
 		       int cmd);
 	int (*dai_link_fixup)(struct snd_soc_pcm_runtime *rtd, struct snd_pcm_hw_params *params);
+	int (*pcm_setup)(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm);
+	void (*pcm_free)(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm);
```
This is confusing for me. All the existing callbacks are self-explanatory, they are traditional 'ALSA' callbacks, so I can picture when they are used.

pcm_setup() is called in sof_dai_load() and pcm_free() is called in sof_dai_unload(). That's far from obvious to me what setup and load really mean... It wouldn't hurt to describe what the 'setup' is supposed to do.

plbossart replied:
there should also be a mention that these callbacks are optional? That's how they are used but it's not clear in the kernel-doc

ranj063 replied:
all of these ops are optional and they don't mention it. We explicitly only call out the mandatory ones and i think that would be simpler,

ujfalusi replied:
It is a bit strange to have these new ops in `sof_ipc_pcm_ops` and call them _only_ from `topology.c`?

plbossart replied:
@ujfalusi I felt the same but it's my understanding that these helpers allocate stuff and prepare the ground for the trigger to operate, so in a way they are companion functions to the trigger and should be in the same ops.

ranj063 replied:
the ops are invoked from topology but since they deal with spcm I think they go better with pcm_ops instead of tplg_ops.

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -215,8 +215,41 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
+						       sizeof(struct snd_sof_widget *),
+						       GFP_KERNEL);
+		if (!pipeline_info->pipeline_list)
+			return -ENOMEM;
```
do we need an unwind loop to free-up the memory that's been allocated earlier in the loop?

ranj063 replied:
yes

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -215,8 +215,41 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
+		pipeline_info = &spcm->stream[stream].pipeline_info;
+
+		/* allocate memory for max number of pipeline IDs */
+		pipeline_info->pipeline_list = kcalloc(ipc4_data->max_num_pipelines,
```
nit-pick: is this max - 1?

If this is the pipelines that are connected, then each FE pipeline cannot be part of the list since they are the head?

ranj063 replied:
not sure I follow @plbossart all pipelines especially FE pipelines must be part of this list.

plbossart replied:
the question is "what does this list represent"?

a) all pipelines reachable from the PCM?
b) all pipelines connected to the first pipeline associated with the PCM.

off-by-one count in the latter case.

ranj063 replied:
a)

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -215,8 +215,41 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
+
+	for_each_pcm_streams(stream) {
+		pipeline_info = &spcm->stream[stream].pipeline_info;
+		kfree(pipeline_info->pipeline_list);
```
do we need to set pipeline_info->pipeline_list = NULL?

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1801,6 +1801,7 @@ static int sof_ipc4_dai_config(struct snd_sof_dev *sdev, struct snd_sof_widget *
 	case SOF_DAI_INTEL_HDA:
 		gtw_attr = ipc4_copier->gtw_attr;
 		gtw_attr->lp_buffer_alloc = pipeline->lp_mode;
+		pipeline->skip_during_fe_trigger = true;
```
I am still in the dark on why we need to skip only HDAudio pipelines?

ranj063 replied:
because they're the only ones left with BE DAI trigger ops

plbossart replied:
Can't we detect this rather than add a flag that duplicates information that can be discovered?

ranj063 replied:
I can think of doing this with a very convoluted check by accessing the snd_soc_dai from the snd_sof_widget using the sname and then checking if the dai->drv-ops are set. But the flag is a lot simpler I think

plbossart replied:
Don't we have a snd_sof_dai -> snd_soc_dai helper? I don't really understand why we don't have a 1:1 mapping.

ranj063 replied:
we don't have a 1:1 mapping, Pierre. I can add this in a follow up maybe

plbossart replied:
That would be good indeed

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -394,29 +396,56 @@ static int sof_free_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dap
+		 * Add the widget's pipe_widget to the list of pipelines to be triggered if not
+		 * already in the list. This will result in the pipelines getting added in the
+		 * order source to sink.
+		 */
```
what happens if we have a split/merge structure, as done in some topologies. Is this ok if the intermediate pipelines sharing the same parent are exposed in no specific order?

Also what happens if we have a fork, e.g. if we play on two interfaces. Which sink should be triggered first?  would that even matter?

ranj063 replied:
> what happens if we have a split/merge structure, as done in some topologies. Is this ok if the intermediate pipelines sharing the same parent are exposed in no specific order?
> 
> Also what happens if we have a fork, e.g. if we play on two interfaces. Which sink should be triggered first? would that even matter?

I really don't think a fork is a special case here. If you consider only one side of the fork, the pipelines are triggered sink->source and that should be enough

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -37,60 +64,96 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
+
+	/*
+	 * IPC4 requires pipelines to be triggered in order starting at the sink and
+	 * walking all the way to the source. So traverse the pipeline_list in the reverse order.
```
does this work with a fork?

ranj063 replied:
with a fork, the order of trigger between the left/right will be indeterministic but it shouldn't matter because what matters is the pipelines for each fork should get triggered from sink->source

plbossart replied:
I think you should capture that in a comment

ranj063 replied:
done

plbossart reviewed (commented):
This looks much better @ranj063, but I still have a number of questions.

ujfalusi commented on sound/soc/sof/sof-audio.h:
```diff
@@ -282,6 +282,16 @@ struct sof_token_info {
 };
 
+/**
+ * struct snd_sof_pipeline_trigger_info - multi pipeline trigger IPC info
```
How this struct is 'multi pipeline trigger **IPC info**'?
It is more like pipeline path or pipeline list.

In any case, I find it increasingly confusing that `snd_sof_widget` is sometimes a widget, module, component or a pipeline.

Probably it is really a snd_sof_pipeline_trigger_path? Certainly not IPC info.

ujfalusi replied:
And this does not necessarily needs to be valid for paths which consists of more than one pipeline (we always have at least 2 pipelines with IPC4, are they also multi-pipeline?)

ranj063 replied:
yes 2 is multi a well.

ranj063 replied:
> In any case, I find it increasingly confusing that `snd_sof_widget` is sometimes a widget, module, component or a pipeline.

a widget is never a pipeline but widget/module/component are all used intrerchageably. There's a pipeline widget which is a widget of type snd_soc_dapm_scheduler.

ranj063 replied:
> How this struct is 'multi pipeline trigger **IPC info**'?

I changed the name to snd_sof_pcm_stream_trigger_info now. hope that makes more sense.

ujfalusi commented on sound/soc/sof/sof-audio.c:
```diff
@@ -358,14 +358,16 @@ sof_prepare_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget
-	if (widget->dobj.private) {
-		err = sof_widget_free(sdev, widget->dobj.private);
+	if (swidget) {
+		err = sof_widget_free(sdev, swidget);
```
Is this change belongs to the topic?

ujfalusi commented on include/sound/sof/ipc4/header.h:
```diff
@@ -185,6 +185,8 @@ enum sof_ipc4_pipeline_state {
 #define SOF_IPC4_GLB_PIPE_STATE_MASK		GENMASK(15, 0)
 #define SOF_IPC4_GLB_PIPE_STATE(x)		((x) << SOF_IPC4_GLB_PIPE_STATE_SHIFT)
 
+#define SOF_IPC4_GLB_PIPE_MULTI			BIT(0)
```
```c
#define SOF_IPC4_GLB_PIPE_STATE_EXT_MULTI	BIT(0)
```
and/or comment that this bit applies to the extension, not to the primary?

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -37,75 +64,96 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
+		return 0;
+
+	/* allocate data for the pipeline ID's and the count of pipelines to be triggered */
+	data = kzalloc(sizeof(u32) * (pipeline_info->count + 1), GFP_KERNEL);
```
```c
data = kzalloc(struct_size(data, pipeline_ids, pipeline_info->count), GFP_KERNEL);
```

ranj063 replied:
this too, can I fix in a fixup?

RanderWang commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -37,75 +64,96 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
-		if (!swidget)
-			continue;
+	/* pause all pipelines */
+	ret = sof_ipc4_set_multi_pipeline_state(sdev, SOF_IPC4_PIPE_PAUSED, data);
```
if all those pipeline are paused first then resumed,  we don't need to paused it again.

ranj063 replied:
yes, it was a conscious choice but maybe it better to avoid that extra IPC. fixed now.

plbossart commented on sound/soc/sof/sof-audio.h:
```diff
@@ -297,6 +307,7 @@ struct snd_sof_pcm_stream {
 	 * active or not while suspending the stream
 	 */
 	bool suspend_ignored;
+	struct snd_sof_pipeline_trigger_info pipeline_info;
```
@ranj063 it looks like @RanderWang is about to add a 'private' field here for IPC-specific stuff in https://github.com/thesofproject/linux/pull/3787/commits/490f6530b6df3233464e7784a5c231fddfc09c54#diff-3aa89cf77fc07290f36500245dad57c983ae8628a8fb4b43804391a01ca4f520R295

How many private things are we going to add here?

ranj063 replied:
@plbossart there's nothing IPC-specific in this struct and I have to insist that this information belongs in snd_sof_pcm_stream. There's just no place else that this information fits in IMHO

ranj063 replied:
The other reason to not move it to a private field @plbossart is the fact that this field is populated in IPC-agnostic sof-audio.c. If I move it inside a private field, I wont be able to do that any more.

bardliao commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -215,8 +215,49 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
+		trigger_info = &spcm->stream[stream].trigger_info;
+		kfree(trigger_info->pipeline_list);
+		trigger_info->pipeline_list = NULL;
+	}
```
Can we just call `sof_ipc4_pcm_free` here?

plbossart replied:
indeed, good suggestion @bardliao

bardliao reviewed (approved):
This PR LGTM and test result looks good

libinyang reviewed (approved):
LGTM

ujfalusi commented on sound/soc/sof/sof-audio.h:
```diff
@@ -282,6 +282,16 @@ struct sof_token_info {
 };
 
+/**
+ * struct snd_sof_pipeline_trigger_info - PCM stream trigger info
```
the struct got renamed to `snd_sof_pcm_stream_trigger_info`

ranj063 replied:
oops can I fix this in a fixup, Peter?

ujfalusi replied:
sure

ujfalusi commented on sound/soc/sof/sof-audio.h:
```diff
@@ -282,6 +282,16 @@ struct sof_token_info {
+ */
+struct snd_sof_pcm_stream_trigger_info {
+	uint32_t count;
+	struct snd_sof_widget **pipeline_list;
```
I'm not good at naming, but would
```
struct snd_sof_pcm_stream_pipeline_list {
	uint32_t count;
	struct snd_sof_widget **pipelines;
};
```
or
```
struct snd_sof_pipelines_in_pcm_stream {
	uint32_t count;
	struct snd_sof_widget **pipeline_list;
};
```
would describe this better?
The reason is that I don't see anything **trigger_info**ish in the struct

ranj063 replied:
@ujfalusi pipeline_list is the list of pipelines to be triggered. Isn't that good enough? And it is only used in the trigger callback.

ujfalusi commented on sound/soc/sof/sof-audio.c:
```diff
@@ -394,29 +395,56 @@ static int sof_free_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dap
+		 * already in the list. This will result in the pipelines getting added in the
+		 * order source to sink.
+		 */
+		for (i = 0; i < trigger_info->count; i++) {
```
Should we be paranoid and protect against `trigger_info->count > ipc4_data->max_num_pipelines`? Probably another member to the trigger_info to store the limit?

ranj063 replied:
no need. if the number of pipelines in tplg were ever greater than max_num_pipelines, we'd have failed long before.

plbossart reviewed (approved):
This is good enough for me if there is a fixup coming to take @bardliao and @ujfalusi 's suggestions into account.

@ujfalusi you get to choose if a new version is needed or if we handle change requests with a fixup.

ujfalusi reviewed (approved):
@ranj063, overall this looks great, amazing actually.
I think it is fine to do fixups for the smaller issues since the functionality is not going to be changed, let's get this merged if everyone is OK with it.

plbossart commented:
merging
